docs/ci_getsock.3
.\"
.\" Copyright 1998 by Daniel Stenberg
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_GETSOCK 3 "11 March 2010"
.SH NAME
ci_getsock \- get socket descriptors to wait on (deprecated)
.SH SYNOPSIS
.nf
#include <ci.h>

int ci_getsock(const ci_channel_t *\fIchannel\fP, ci_socket_t *\fIsocks\fP,
                 int \fInumsocks\fP);
.fi
.SH DESCRIPTION
The
.B ci_getsock
function retrieves the set of socket descriptors which the calling
application should wait on for reading and/or writing for the
processing of name service queries pending on the name service channel
identified by
.IR channel .
Socket descriptors will be set in the socket descriptor array pointed to by
\fIsocks\fP.
\fInumsocks\fP is the size of the given array in number of ints.

This function can only return information up to 16 sockets. If more are
in use, they are simply not reported back.
.SH RETURN VALUES
\fBci_getsock\fP returns a bitmask for what actions to wait for on the
different sockets. The ci.h header file provides these convenience macros to
extract the information appropriately:

.nf
#define CI_GETSOCK_MAXNUM 16 /* ci_getsock() can return info about
                                  this many sockets */
#define CI_GETSOCK_READABLE(bits,num) (bits & (1<< (num)))
#define CI_GETSOCK_WRITABLE(bits,num) (bits & (1 << ((num) + \
                                         CI_GETSOCK_MAXNUM)))
.fi
.SH NOTES
This function was added in c-ci 1.3.1 and deprecated in c-ci 1.20.0 due to
the implementation of \fBCI_OPT_MAX_UDP_QUERIES\fP which makes it likely to
exceed 16 open file descriptors.

It is recommended to use \fBCI_OPT_EVENT_THREAD\fP passed to
\fIci_init_options(3)\fP or to use socket state callbacks
(\fBCI_OPT_SOCK_STATE_CB\fP) registered via \fBci_init_options(3)\fP.
.SH SEE ALSO
.BR ci_init_options (3),
.BR ci_timeout (3),
.BR ci_fds (3),
.BR ci_process (3)


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/legacy/ci_getsock.c
/* MIT License
 *
 * Copyright (c) 2005 Daniel Stenberg
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ci_private.h"

int ci_getsock(const ci_channel_t *channel, ci_socket_t *socks,
                 int numsocks) /* size of the 'socks' array */
{
  ci_slist_node_t *snode;
  size_t             sockindex = 0;
  unsigned int       bitmap    = 0;
  unsigned int       setbits   = 0xffffffff;

  /* Are there any active queries? */
  size_t             active_queries;

  if (channel == NULL || numsocks <= 0) {
    return 0;
  }

  ci_channel_lock(channel);

  active_queries = ci_llist_len(channel->all_queries);

  for (snode = ci_slist_node_first(channel->servers); snode != NULL;
       snode = ci_slist_node_next(snode)) {
    ci_server_t     *server = ci_slist_node_val(snode);
    ci_llist_node_t *node;

    for (node = ci_llist_node_first(server->connections); node != NULL;
         node = ci_llist_node_next(node)) {
      const ci_conn_t *conn = ci_llist_node_val(node);

      if (sockindex >= (size_t)numsocks || sockindex >= CI_GETSOCK_MAXNUM) {
        break;
      }

      /* We only need to register interest in UDP sockets if we have
       * outstanding queries.
       */
      if (!active_queries && !(conn->flags & CI_CONN_FLAG_TCP)) {
        continue;
      }

      socks[sockindex] = conn->fd;

      if (active_queries || conn->flags & CI_CONN_FLAG_TCP) {
        bitmap |= CI_GETSOCK_READABLE(setbits, sockindex);
      }

      if (conn->state_flags & CI_CONN_STATE_WRITE) {
        /* then the tcp socket is also writable! */
        bitmap |= CI_GETSOCK_WRITABLE(setbits, sockindex);
      }

      sockindex++;
    }
  }

  ci_channel_unlock(channel);
  return (int)bitmap;
}
